MacroStageXY improvements

dstoychev opened the pull request:
These are the changes necessary to make the MacroStageXY draw correctly regardless of the canvas size. While at it, I fixed the way the scalebar and cross hairs are drawn, as they seemed wrong. There also some other small bugfixes, like not drawing the corner labels if the soft limits coincide with the hard limits.

Addresses #585.

Here are before and after images, respectively:

![before](https://user-images.githubusercontent.com/56877735/80346972-8c95f880-8863-11ea-8a0e-f7ccd9c338d1.png)

![after](https://user-images.githubusercontent.com/56877735/80346980-90c21600-8863-11ea-887b-f9a270d0a219.png)

iandobbie commented:
Yes, looks better. I had previously struggled with weird xy scaling in the macrostage xy view. Seems you have now fixed that. Well done!

iandobbie commented:
I have just pulled this in with my virtual stage branch and there are a few issues. 

<img width="877" alt="macrostage-updates" src="https://user-images.githubusercontent.com/5451909/80481543-934b6b00-894a-11ea-8f57-b5303a572646.png">

1) The cross hairs have got too small to see in with this (admittedly tiny) virtual stage

2) Why are there so few tick marks on the main Z axis?

3) Also is there not a bit too much space around the stage and between the text, surely the stage XY map could be 30% bigger and the text below it squashed up a bit

4) I think this is a separate issue. Recenter leaves the 3rd Z axis displaying the wrong region. Shifting the stage corrects the displayed region

iandobbie commented:
I just spoent quite some time chasing the missing image position in the macroStageXY view and just can't find out why it is gone.

dstoychev commented:
> 1. The cross hairs have got too small to see in with this (admittedly tiny) virtual stage

That's easy to fix, I will address it in the next commit.

> 2. Why are there so few tick marks on the main Z axis?

The tick marks are drawn every 1000um for the minor and every 5000um for the major. I assume your stage is 10mm by 10mm, so everything seems to be in order.

> 3. Also is there not a bit too much space around the stage and between the text, surely the stage XY map could be 30% bigger and the text below it squashed up a bit

This is a bit more difficult to solve. I have a couple of ideas and I will see what works the best.

> 4. I think this is a separate issue. Recenter leaves the 3rd Z axis displaying the wrong region. Shifting the stage corrects the displayed region.

Admittedly, I have not tried the recentre functionality properly because I have only one (dummy) stage. Can you tell me more about your setup so that I can try to replicate the issue? I also assume your last comment is related to this?

iandobbie commented:
Turns out issue 1 was the fact that my objectives all have the colour (1,1,1) so the location marks are drawn in white! I think you must have added code to change the square colour as well as the reachable stage positions.

iandobbie commented:
I think this works and we should pull it. If no objections are expressed I will pull it later today. Then open a new issue for point 4 above

dstoychev commented:
There are some extreme cases where the drawing doesn't look very good because I've used hard-coded values for the margins and some other measures. This is especially noticeable when the width is small, as the proportions become more pronounced. There are also some other issues that I wanted to address, but I am quite happy to make them on top of this, as a separate issue.

Regarding the objective colour, that's not something I have changed. The code uses the colour returned by the ObjectiveHandler, which defaults to (1.0, 1.0, 1.0). Even then, all objectives with magnification of 100x or higher are colour coded as white, so maybe it's not a bad idea to revise how the location square is drawn. Perhaps always drawing with the same colour would be better.

carandraug commented:
> Here are before and after images, respectively:

These before and after images are for our default stages but the issue #585 is about "different shaped stages" being badly drawn so I can't see the fix from them. I have asked @iandobbie for more details on the original issue.

> There are some extreme cases where the drawing doesn't look very good because I've used hard-coded values for the margins and some other measures. This is especially noticeable when the width is small, as the proportions become more pronounced. There are also some other issues that I wanted to address, but I am quite happy to make them on top of this, as a separate issue.

Regarding this hard coded values, can they be named? The code for this is already quite complicated to follow, having a name for them would make it a bit easier to follow. I have no clue where this numbers come from `for i, (dx, dy) in enumerate([(400, -1000), (0, 800)])`

carandraug commented:
I tried with a bit more different stage sizes (see original issue for a patch on how to create XY stages of various sizes). I have found that the caption under the text under the macro stage with the current coordinates, overlap each other for some sizes, e.g.:

![danny-25000x50000](https://user-images.githubusercontent.com/916140/80928420-30b7fc00-8d9c-11ea-8954-0d2f5133624e.png)

Also, the new tick marks between the current coordinates and the macro stage itself seems a new feature, it should at least be on its own separate commit.

carandraug commented:
> Also, the new tick marks between the current coordinates and the macro stage itself seems a new feature, it should at least be on its own separate commit.

Actually, I was just testing against some code from a few years ago and turns out that the tick marks at the bottom of the macro XY stage are not new. I have been going through the logs and found that it stopped working with 53d5a351b890. Still, we should individual commits to address individual issues if possible.

iandobbie commented:
Yes, I could have told you it used to be there, I think Danny just fixed our broken code.

dstoychev commented:
I addressed all the feedback and I moved the coordinate labels of the MacroStageXY to the top, as asked in #594. Here is an image that shows how the macro stage menu looks on my branch (a bit outdated), as well as a short animation that demonstrates the fitting:

![output](https://user-images.githubusercontent.com/56877735/81592975-b0565400-93b6-11ea-9bd6-89cbf08623ac.png)

![output](https://user-images.githubusercontent.com/56877735/81593073-d11ea980-93b6-11ea-8283-0d996b52dd8c.gif)

iandobbie commented:
I think we should test in master and pull it in so long as there are no obvious issues. It is obviously so much better than the repeated hacks to try and solve specific stage shapes done previously, and bring back our tick marks etc....

carandraug commented:
Here's some comments reviewing this:

* while the macro stage is supposed to now expand to the space available, the font should not increase/decrease in size.  I think this can be addressed by not having the coordinates text in openGL to start with.  This might be easier to do if you start by doing it first (which fixes #594) since then there will be less code here to fix.

* After changing the stage safeties, the text showing the current coordinates stops updating.

* `MacroStageXY.calculateMeasures` declares multiple attributes.  All attributes should be declared inside the body of `__init__` (even if those methods that declare new attributes are now called from `__init__`).

* Of the attributes mentioned above, most of them are only used by `OnPaint` which starts by calling `calculateMeasures`.  That suggests that those attributes should be limited to that function (maybe a return value of `calculateMeasures`).

* Also following on the previous point, `softlimit_label_line_height` and `margin` are only used within `calculareMeasures`, the function that computes them, so they should really not be an attribute of the class.

* It's a bit odd how on the course of `calculateMeasures` (70 lines long!) the value of maxX/Y and minX/Y keeps changing as the measurements of the different parts get compute.  Considering the above points, would it make sense to have a `calculateScalebarMeasures` function and similar, each being called when those measures are needed in `OnPaint`?  Or maybe even have a `PaintScalerbar`, `PaintCoorLabels`, etc, which `OnPaint` calls?

* Your message says that this changes made the corner labels (text showing the coordinates of the corners) only appear if the soft limits are different fro the hard limits as was meant to.  However, I never get the corner labels at all.  Also, since we use the non-standard lower limits at the bottom-right (instead of bottom-left), maybe we should always show those coordinates (maybe this is a separate issue at all).

* Please keep lines at a maximum of 79 (inclusive) characters if possible.

* Is there any reason to replace the custom projection matrix with glOrtho?  From the documentation, I'm guessing these are equivalent. If so, can the commit message be specific about the reason, something like 'MacroStageXY: replace custom projection matrix with equivalent glOrtho'? (I can do this ameding of commit messages myself later since there's other things I would like to add such as references to issue number).

dstoychev commented:
Thanks for the feedback, greatly appreciated! I want to address all points, hopefully it's not too confusing:

>the font should not increase/decrease in size

I am not sure that I agree. I intentionally made the text to scale because I figured if the canvas is going to be made big, the text should be big as well. It doesn't seem right to have the stage and scale bar large, while the text is small.

>After changing the stage safeties, the text showing the current coordinates stops updating.

Found the cause -- a bug in the string formatting. Fixed in 87b753f.

>All attributes should be declared inside the body of ``__init__``

Not needed any more since c9990cd.

>attributes should be limited to [the ``calculateMeasures``] function

Fixed in c9990cd.

>[``softlimit_label_line_height`` and ``margin``] should really not be an attribute of the class

Fixed in 065dcc7.

>[W]ould it make sense to have a calculateScalebarMeasures function and similar, each being called when those measures are needed in OnPaint? Or maybe even have a PaintScalerbar, PaintCoorLabels, etc, which OnPaint calls?

I think the problem with this is that the drawing parameters (previous in ``calculateMeasures()``, now in ``calculateDrawingParams()``) need to be calculated for the orthographic projection before the actual drawing takes place. Now, this begs the question why are we setting the projection in the paint event when it needs to be done only once. I think there is definitely room for imporvement there.

>I never get the corner labels at all

This is related to the issue of coordinates not updating, which I fixed in 87b753f. The labels are only drawn if they do not coincide with the hard limits. If you change the safeties after the fix the labels should appear.

>Please keep lines at a maximum of 79 (inclusive) characters if possible

Fixed in 6faa9c8.

>Is there any reason to replace the custom projection matrix with glOrtho?

My reasoning was that there is no need to have custom code for something that could be done with the standard OpenGL API (even if deprecated). I find it confusing and unnecessary.

carandraug commented:
>> the font should not increase/decrease in size
>
> I am not sure that I agree. I intentionally made the text to scale
> because I figured if the canvas is going to be made big, the text
> should be big as well. It doesn't seem right to have the stage and
> scale bar large, while the text is small.

It's not about the canvas being made big but not the text. It's about the canvas taking up the space available while the text staying at the size the user configured it to be.  The stage and the scale bar can be made larger to show more detail but making the font larger does not add more.  And it works the other way around.  The stage and scale bar might need to be made smaller if the canvas has less space but the font must remain at a size that is readable (which also puts a lower limit on the size).

Obviously this assumes that the font is set to something that the user set it to instead of the current hardcoded size 18 but that's a different issue.  If we move the coordinates text to wx as proposed in #594 that's what will happen for that text at least.

>> After changing the stage safeties, the text showing the current
>> coordinates stops updating.
>
> Found the cause -- a bug in the string formatting. Fixed in 87b753f.

We used the old, printf C-style, % for formatting everywhere else. Can you use it, for consistency at least?

> Now, this begs the question why are we setting the projection in the
> paint event when it needs to be done only once. I think there is
> definitely room for imporvement there.

If you think that the fix for the drawing of the different shaped stages will be simpler if you first separate the computing of drawing parameters from the actual drawing, please do that first.

From the code, it seems that the drawing parameters are dependent on the window client size, stage hard limits, and objective offsets.  If that's the case, since the hard limits and objective offsets do not change, I guess you could just recompute them when the window size changes.

>> Please keep lines at a maximum of 79 (inclusive) characters if possible
>
> Fixed in 6faa9c8.

Well, technically under 80 characters.  We need to define some sort of style since we are often not consistent even across the same file.

>> Is there any reason to replace the custom projection matrix with glOrtho?
>
> My reasoning was that there is no need to have custom code for
> something that could be done with the standard OpenGL API (even if
> deprecated). I find it confusing and unnecessary.

That's a fine reason.  It was simply not stated anywhere and I had to check OpenGL documentation to guess that the two were equivalent. Would be nice to have that sort of thing on the commit message. If that is not stated somewhere, if someone wants to revert it in the future they might think the change was to fix an issue and will have to spend time figuring out what that non-existent issue was.
